orchestrator/careplancontributor/service.go

```diff
@@ -184,35 +184,38 @@ func (s Service) handleProxyToFHIR(writer http.ResponseWriter, request *http.Req
 	if err != nil {
 		return err
 	}
-	isValidRequester := false
+
+	isValid, err := validateRequester(careTeams, principal)
+	if err != nil {
+		return err
+	}
+
+	if !isValid {
+		return coolfhir.NewErrorWithCode("requester does not have access to resource", http.StatusForbidden)
+	}
+	fhirProxy := coolfhir.NewProxy(log.Logger, s.fhirURL, basePath+"/fhir", s.transport)
+	fhirProxy.ServeHTTP(writer, request)
+	return nil
+}
+
+// validateRequester loops through each Participant of each CareTeam present in the CarePlan, trying to match it to the Principal
+// if a match is found, validate the time period of the Participant
+func validateRequester(careTeams []fhir.CareTeam, principal auth.Principal) (bool, error) {
 	for _, careTeam := range careTeams {
 		for _, participant := range careTeam.Participant {
 			for _, identifier := range principal.Organization.Identifier {
 				if coolfhir.IdentifierEquals(participant.OnBehalfOf.Identifier, &identifier) {
 					// Member must have start date, this date must be in the past, and if there is an end date then it must be in the future
-					err = coolfhir.ValidateCareTeamParticipantPeriod(participant, time.Now())
+					err := coolfhir.ValidateCareTeamParticipantPeriod(participant, time.Now())
 					if err != nil {
-						return err
+						return false, err
 					}
```

Comment:
There's an edge case (although currently not possible now); if the principal's org has multiple identifiers, of which multiple occur in the careTeam, if the first match has an expired period, it invalidates the membership. But, a there could be a subsequent membership that's not expired. Although it would be an extreme edge case, I think it would simplify things if you just return `true` when there's a matching identifier with a non-expired period.

I think this roots from the fact that `ValidateCareTeamParticipantPeriod` should return a bool (whether or not expired) AND an error (invalid object). In case of an error, you can just log it (WARN) and ignore the entry, and try the next identifier.

Reply:
That sounds reasonable, I hadn't considered that edge case so thanks for bringing it to my attention!